tools/terraform/provider/teams_resource.go

```diff
@@ -0,0 +1,288 @@
+	return &teamsResource{
+		client: nil,
+	}
+}
+
+// teamsResource is the resource implementation.
+type teamsResource struct {
+	client *fleetdm_client.FleetDMClient
+}
+
+// Configure adds the provider configured client to the resource.
+func (r *teamsResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*fleetdm_client.FleetDMClient)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *FleetDMClient, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	r.client = client
+}
+
+// Metadata returns the resource type name.
+func (r *teamsResource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_teams"
+}
+
+// Schema defines the schema for the resource.
+func (r *teamsResource) Schema(ctx context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = TeamResourceSchema(ctx)
+}
+
+// teamModelToTF is a handy function to take the results of the API call
+// to the FleetDM API and convert it to the Terraform state.
+func teamModelToTF(ctx context.Context, tm *fleetdm_client.TeamGetResponse, tf *TeamModel) error {
+	tf.Id = types.Int64Value(tm.Team.ID)
+	tf.Name = types.StringValue(tm.Team.Name)
+	tf.Description = types.StringValue(tm.Team.Description)
+	tf.Secrets = basetypes.NewListNull(NewSecretsValueNull().Type(ctx))
+	// Re-marshal agent_options into a string so TF can store it sanely.
+	aoBytes, err := json.Marshal(tm.Team.AgentOptions)
+	if err != nil {
+		return fmt.Errorf("failed to re-marshal agent options: %w", err)
+	}
+	tf.AgentOptions = types.StringValue(string(aoBytes))
+
+	return nil
+}
+
+// Create creates the resource from the plan and sets the initial Terraform state.
+func (r *teamsResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Retrieve values from plan
+	var plan TeamModel
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	newTeam, err := r.client.CreateTeam(plan.Name.ValueString(), plan.Description.ValueString())
+	if err != nil {
+		resp.Diagnostics.Append(diag.NewErrorDiagnostic(
+			"Failed to create team",
+			fmt.Sprintf("Failed to create team: %s", err)))
+		return
+	}
+
+	if !plan.AgentOptions.IsNull() && !plan.AgentOptions.IsUnknown() {
+		aoPlan := plan.AgentOptions.ValueString()
+		if aoPlan != "" {
+			newTeam, err = r.client.UpdateAgentOptions(newTeam.Team.ID, aoPlan)
+			if err != nil {
+				resp.Diagnostics.Append(diag.NewErrorDiagnostic(
+					"failed to create agent options",
+					fmt.Sprintf("failed to save agent options: %s", err)))
+				_ = r.client.DeleteTeam(plan.Id.ValueInt64()) // Problematic. :-/
```

Comment:
Is there a proper way to surface this in terraform? Like some warning that cleanup could not be completed successfully?

Reply:
I added a more informative comment as well as added another error diagnostic.